src/wayland/idle_notify/mod.rs

```diff
@@ -0,0 +1,345 @@
+    fn idle_notifier_state(&mut self) -> &mut IdleNotifierState<Self>;
+}
+
+/// User data of [ExtIdleNotificationV1] resource
+#[derive(Debug)]
+pub struct IdleNotificationUserData {
+    seat: WlSeat,
+    is_idle: AtomicBool,
+    timeout: Duration,
+    timer_token: Mutex<Option<RegistrationToken>>,
+}
+
+impl IdleNotificationUserData {
+    fn take_timer_token(&self) -> Option<RegistrationToken> {
+        self.timer_token.lock().unwrap().take()
+    }
+
+    fn set_timer_token(&self, idle: Option<RegistrationToken>) {
+        *self.timer_token.lock().unwrap() = idle;
+    }
+
+    fn set_idle(&self, idle: bool) {
+        self.is_idle.store(idle, atomic::Ordering::Release);
+    }
+
+    fn is_idle(&self) -> bool {
+        self.is_idle.load(atomic::Ordering::Acquire)
+    }
+}
+
+type SeatObjectId = ObjectId;
+type NotificationObjectId = ObjectId;
+
+type SeatMap<T> = HashMap<SeatObjectId, T>;
+type NotificationMap<T> = HashMap<NotificationObjectId, T>;
+
+/// State of ext-idle-notify module
+#[derive(Debug)]
+pub struct IdleNotifierState<D> {
+    global: GlobalId,
+    notifications: SeatMap<NotificationMap<ExtIdleNotificationV1>>,
+    loop_handle: LoopHandle<'static, D>,
+    is_inhibited: bool,
+}
+
+impl<D: IdleNotifierHandler> IdleNotifierState<D> {
+    /// Create new [`ExtIdleNotifierV1`] global.
+    pub fn new(display: &DisplayHandle, loop_handle: LoopHandle<'static, D>) -> Self
+    where
+        D: GlobalDispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotificationV1, IdleNotificationUserData>,
+        D: IdleNotifierHandler,
+        D: 'static,
+    {
+        let global = display.create_global::<D, ExtIdleNotifierV1, _>(1, ());
+        Self {
+            global,
+            notifications: HashMap::new(),
+            loop_handle,
+            is_inhibited: false,
+        }
+    }
+
+    /// Inhibit entering idle state, eg. by wp idle inhibit protocol
+    pub fn set_is_inhibited(&mut self, is_inhibited: bool) {
+        if self.is_inhibited == is_inhibited {
+            return;
+        }
+
+        self.is_inhibited = is_inhibited;
+
+        for notification in self.notifications() {
+            if is_inhibited {
+                let data = notification.data::<IdleNotificationUserData>().unwrap();
+
+                if data.is_idle() {
+                    notification.resumed();
+                }
+
+                if let Some(token) = data.take_timer_token() {
+                    self.loop_handle.remove(token);
+                }
+            } else {
+                self.reinsert_timer(notification);
+            }
+        }
+    }
+
+    /// Is idle state inhibited, eg. by wp idle inhibit protocol
+    pub fn is_inhibited(&mut self) -> bool {
+        self.is_inhibited
+    }
+
+    /// Should be called whenever activity occurs on a seat, eg. mouse/keyboard input
+    pub fn notify_activity(&mut self, seat: &WlSeat) {
```

Comment:
Shouldn't this accept a `Seat`? Is it even possible to get all WlSeats out of a Seat?